htpasswd.go

```diff
@@ -177,30 +216,66 @@ func (h *HtpasswdFile) ReloadCheck() error {
 func (h *HtpasswdFile) Validate(user string, password string) bool {
 	_ = h.ReloadCheck()
 
+	hash := sha256.New()
+	// hash.Write can never fail
+	_, _ = hash.Write([]byte(user))
+	_, _ = hash.Write([]byte(":"))
+	_, _ = hash.Write([]byte(password))
+
 	h.mutex.Lock()
-	realPassword, exists := h.Users[user]
+	// avoid race conditions with cache replacements
+	cache := h.cache
+	realPassword, exists := h.users[user]
+	entry, cacheExists := h.cache[user]
 	h.mutex.Unlock()
 
 	if !exists {
 		return false
 	}
 
+	if cacheExists && subtle.ConstantTimeCompare(entry.verifier, hash.Sum(nil)) == 1 {
+		h.mutex.Lock()
+		// repurpose mutex to prevent concurrent cache updates
+		// extend cache entry
+		cache[user] = cacheEntry{
+			verifier: entry.verifier,
+			expiry:   time.Now().Add(PasswordCacheDuration),
+		}
+		h.mutex.Unlock()
+		return true
+	}
+
```

Comment:
Does "hits a cache" need to be constant time? Surely we only want constant time verification when checking the password.

Reply:
Good question, I think that an attacker could potentially learn whether a part of the user:password-hash has matched. It's probably very hard to exploit as a lot of offline bruteforcing would be necessary, to generate hashes which a fixed prefix incrementally. Does using a constant time comparison here cause any problem?